.sandbox/provision/lifecycle-asg/common.sh

```diff
@@ -0,0 +1,101 @@
+        local instance_id=""
+        instance_id="$(aws autoscaling describe-auto-scaling-groups --auto-scaling-group-name "$ASG_NAME" --no-paginate --query AutoScalingGroups[0].Instances[0].InstanceId --output text)"
+
+        # detach-instances is an atomic operation and a single instance can not be deatched twice. 
+        # This mechanism is used to prevent race condition between two or more claims that happening at the same time. The second claim would fail.
+        aws autoscaling detach-instances --instance-ids "$instance_id" --auto-scaling-group-name "$ASG_NAME" --no-should-decrement-desired-capacity >/dev/null
+
+        # Two tags are created for the detached instance
+        # - Sandbox, this is for referencing purpose and should not be used as an ID during looking up.
+        # - SandboxID, this is the actual identification of the EC2 instance.
+        # An EC2 instance is deemed as leaked if:
+        # - SandboxID tag is missing
+        # - Some ASG flagging tag exists, e.g. sandbox-asg=true
+        # - The instance is detached from ASG.
+        aws ec2 create-tags --resources "$instance_id" --tags Key=$SANDBOX_NAME_TAG,Value="$SANDBOX_NAME" --tags Key=$SANDBOX_ID_TAG,Value="$SANDBOX_ID"
+
+        echo "$instance_id"
+    else 
+        jq -cMr '.Reservations[0].Instances[0].InstanceId' <<< "$instance_with_sandbox_id"
+    fi
+}
+
+function attach_volume_if_needed() {
+    local volume_id="$1"
+    local instance_id="$2"
+    local volume=""
+
+    aws ec2 wait volume-available --volume-ids "$volume_id"
+    volume="$(aws ec2 describe-volumes --volume-id "$volume_id")"
+    [[ "$(jq -cMr ".Volumes[0].Attachments | length" <<< "$volume")" -gt 0 ]] || {
+        aws ec2 attach-volume --volume-id "$volume_id" --instance-id "$instance_id" --device "$DEVICE_NAME"
+    }
+}
+
+# retrieve_password INSTANCE_ID KEY_FILE
+function retrieve_password() {
+    local instance_id="$1"
+    local ec2_ssh_key_file="$2"
+    local password_data=""
+
+    password_data="$(aws ec2 get-password-data --instance-id "$instance_id" --query 'PasswordData' --output text)"
+
+    local retry_count=0
+    while [[ "$retry_count" -lt "$MAX_RETRIES" ]]; do
+        password_data="$(aws ec2 get-password-data --instance-id "$instance_id" --query 'PasswordData' --output text)"
+        [[ -z "$password_data" ]] || break
+        retry_count=$((retry_count + 1))
+        sleep 10
+    done
+
+    echo "$password_data" | base64 -d  | openssl pkeyutl  -decrypt -inkey "$ec2_ssh_key_file"
+}
+
+# validate_asg ASG_NAME
+function validate_asg() {
+    local auto_scaling_group_name="$1"
+    local result=""
+
+    result="$(aws autoscaling describe-auto-scaling-groups --auto-scaling-group-names "$auto_scaling_group_name")"
+    [[ $(jq -cMr '.AutoScalingGroups | length' <<< "$result") -gt 0 ]] || fatal "Invalid auto scaling group: $auto_scaling_group_name"
```

Comment:
For better performance and less resource usage, use:

```sh
result="$(aws ... | jq -cMr ...)"
```
as `result` is not used for other purpose.

Reply:
Also, `jq` may be omitted if the aws command fails if nothing found.

Reply:
AWS command would not fail if nothing found, it will return empty array. It fails only something goes wrong with the syntax or API communication error. 

In case of AWS command failure, as I mentioned in another comment, just let it crash. Output from the AWS cli should be sufficient for user to figure what happened.

Reply:
TBH I don't like this practice to leave to the users debugging from the complex logs - this is not the right way to build a user-facing product. But due to the limited effort, I'm OK with `set -ex` for now. However that doesn't answer my question, isn't it OK if AWS cli returns empty array because the next line `jq ... | length` will anyway catch that? My point of the original comment is we can just pipe like `aws ... | jq` without saving an intermediate data in `result` which is never used later.

Reply:
Let's discuss this in our catch up and decide what should be the agreed way to deal with this.